### ISSUE TITLE
Getting ground truth info from the simulator

### ISSUE THREAD

**Issue by mariogpascual** [upkie/vulp#41]
As we talked in tasts-robots/upkie#104, could you give more context on how to obtain ground truth information from the simulator to extend Upkie's observation dictionary? Particularly the height of a torso-attached point. Thank you!

**Comment by stephane-caron** [comment]
What differentiates simulation data from other sensory data is that we collect it inside the interface (as opposed to e.g. CPU temperature or joystick inputs which are read in the observer pipeline). This is not a blocker though, we can proceed similarly to the way IMU data is currently collected.

## Sketch on how to proceed

I see mainly six steps:

1. In ``BulletInterface::cycle``, call a new function to read extra data from the simulator:

```
  read_joint_sensors();
  read_imu_data(imu_data_, bullet_, robot_, imu_link_index_, params_.dt);
  read_extra_data(...);  // new
```

2. This data could have a custom ``ExtraData`` type similar to the ``ImuData`` implemented in ``Interface`` (for the pi3hat interface it would default to an empty class)
3. Add a ``latest_extra_`` attribute to the ``Spine`` class in ``Spine.h``:

```c++
  std::vector<actuation::moteus::ServoReply> latest_replies_;

  //! Latest IMU data. It is copied and thread-safe.
  actuation::ImuData latest_imu_data_;

  //! Latest extra (e.g. simulation) data.
  actuation::ExtraData latest_extra_data_;
```

4. Collect this new observation upon receiving a new actuation output, in ``Spine.cpp``:

```c++
  // 3. Wait for the result of the last query and copy it
  if (actuation_output_.valid()) {
    // ...
    latest_imu_data_ = actuation_.imu_data();
    latest_extra_data_ = actuation_.extra_data(); // new
  }
```

6. Convert this output to a dictionary when cycling the spine, in ``Spine.cpp``:
 
```c++
void Spine::cycle_actuation() {
  try {
    // 1. Observation
    Dictionary& observation = dict_("observation");
    observation::observe_time(observation);
    observation::observe_servos(observation, actuation_.servo_joint_map(),
                                latest_replies_);
    observation::observe_imu(observation, latest_imu_data_);
    observation::observe_extra(observation, latest_extra_data_);
```

With this pipeline, the spine can forward data computed within Bullet to the output observation dictionary.

**Comment by stephane-caron** [comment]
This feature has been shipped [in v2.5.0](https://github.com/upkie/vulp/releases/tag/v2.5.0):

> - BulletInterface: Report the base state that can be used for training RL agents.

Here is an example reading the base height (the base frame is attached to Upkie's torso) and tuning some gain (using [mpackview](https://pypi.org/project/mpackview/) and [valmix](https://pypi.org/project/valmix/)):

![image](https://github.com/upkie/vulp/assets/1189580/ac2f5f35-ff86-41ba-8675-a95c0860fa50)

Hoping this helps!
